addons/mail/models/mail_activity.py

```diff
@@ -149,32 +205,44 @@ def _compute_res_name(self):
 
     @api.depends('date_deadline')
     def _compute_state(self):
-        today_default = date.today()
-
         for record in self.filtered(lambda activity: activity.date_deadline):
-            today = today_default
             tz = record.user_id.sudo().tz
-            if tz:
-                today_utc = pytz.UTC.localize(datetime.utcnow())
-                today_tz = today_utc.astimezone(pytz.timezone(tz))
-                today = date(year=today_tz.year, month=today_tz.month, day=today_tz.day)
-
-            date_deadline = fields.Date.from_string(record.date_deadline)
-            diff = (date_deadline - today)
-            if diff.days == 0:
-                record.state = 'today'
-            elif diff.days < 0:
-                record.state = 'overdue'
-            else:
-                record.state = 'planned'
+            date_deadline = record.date_deadline
+            record.state = self._compute_state_from_date(date_deadline, tz)
+
+    @api.model
+    def _compute_state_from_date(self, date_deadline, tz=False):
+        date_deadline = fields.Date.from_string(date_deadline)
+        today_default = date.today()
+        today = today_default
+        if tz:
+            today_utc = pytz.UTC.localize(datetime.utcnow())
+            today_tz = today_utc.astimezone(pytz.timezone(tz))
+            today = date(year=today_tz.year, month=today_tz.month, day=today_tz.day)
+        diff = (date_deadline - today)
+        if diff.days == 0:
+            return 'today'
+        elif diff.days < 0:
+            return 'overdue'
+        else:
+            return 'planned'
 
     @api.onchange('activity_type_id')
     def _onchange_activity_type_id(self):
         if self.activity_type_id:
             self.summary = self.activity_type_id.summary
             # Date.context_today is correct because date_deadline is a Date and is meant to be
             # expressed in user TZ
-            self.date_deadline = fields.Date.context_today(self) + timedelta(days=self.activity_type_id.days)
+            base = fields.Date.context_today(self)
+            if self.activity_type_id.delay_from == 'previous_activity' and 'activity_previous_deadline' in self.env.context:
```

Comment:
I think you could use something like
``` python
self.date_deadline = base + relativedelta(**{self.activity_type_id.delay_unit: self.activity_type_id.delay_count}) 
```
instead of converting the weeks into days and making a specific case for months. It's a little more straightforward ! ;)

Reply:
relativedelta is not working with months, that's why (28 to 31 days)